packages/modular-scripts/src/esbuild-scripts/plugins/workerFactoryPlugin.ts

```diff
@@ -0,0 +1,141 @@
+
+        // Build the worker file with the same format, target and definitions of the bundle
+        try {
+          const result = await esbuild.build({
+            format: build.initialOptions.format,
+            target: build.initialOptions.target,
+            define: build.initialOptions.define,
+            minify: build.initialOptions.minify,
+            entryPoints: [args.path],
+            plugins: [createExtensionAllowlistPlugin()],
+            bundle: true,
+            write: false,
+          });
+
+          // Store the file in the build cache for later use, since we need to emit a file and trampoline it transparently to the user
+          workerBuildCache.set(relativePath, result);
+
+          // Trampoline the worker within the bundle, to avoid CORS errors
+          return {
+            contents: `
+                  // Web worker bundled by worker-factory-plugin, mimicking the Worker constructor
+                  import workerUrl from 'worker-url:${relativePath}';
+                  
+                  const workerPath = new URL(workerUrl, import.meta.url);
+                  const importSrc = 'import "' + workerPath + '";';
+
+                  const blob = new Blob([importSrc], {
+                    type: "text/javascript",
+                  });
+
+                  export default class {
+                    constructor() {
+                      return new Worker(URL.createObjectURL(blob), { type: "module" });
+                    }
+                  }
+                `,
+          };
+        } catch (e) {
+          console.error('Error building worker script:', e);
+        }
+      });
+
+      build.onResolve({ filter: /worker-url:.*/ }, (args) => {
+        return {
+          path: args.path.slice('worker-url:'.length),
+          namespace: 'worker-url',
+        };
+      });
+
+      build.onLoad({ filter: /.*/, namespace: 'worker-url' }, (args) => {
+        const result = workerBuildCache.get(args.path);
+        if (result) {
+          const { outputFiles } = result;
+          if (outputFiles?.length === 1) {
+            const outputFile = outputFiles[0];
+            return {
+              contents: outputFile.contents,
+              loader: 'file',
+            };
+          } else {
+            throw new Error(`Could not read output files`);
+          }
+        } else {
+          throw new Error(`Could not find result for ${args.path}`);
+        }
+      });
+    },
+  };
+
+  return plugin;
+}
+
+function createExtensionAllowlistPlugin(
+  allowedExtensions = ['.js', '.jsx', '.ts', '.tsx'],
+): esbuild.Plugin {
+  return {
+    name: 'worker-factory-plugin',
+    setup(build) {
+      // No lookbehind in Go regexp; need to look at all the files and do the check manually.
+      build.onResolve({ filter: /.*/ }, (args) => {
+        // Extract the extension; if not in the allow list, throw.
+        const extension = path.extname(args.path);
```

Comment:
Should this be an `onLoad` since imports normally follow module convention naming and therefore don't have any extensions.

Reply:
> Should this be an onLoad since imports normally follow module convention naming and therefore don't have any extensions.

The idea was that implicit imports are always allowed, because they fall in our allow list in the default case. But now that this plugin is split and can be reused, I think you're right. The only con is that we lose the `importer` in the error message (unless we pass it in the plugin data, but tbh from a performance perspective I'm not terribly happy to run `onResolve` and `onLoad` for all the files in the worker bundle).

I changed it to use `onLoad` and retouched the error message a bit, lmk what you think.

Reply:
So rather than throw an error - keep it as `onResolve` and return an error message as part of the resolve hook. This will implicitly make the error appear properly in esbuilds output which is the best of both worlds! 🔥

Reply:
@LukeSheard done!